lib/dao.js

```diff
@@ -2356,6 +2559,200 @@ DataAccessObject.prototype.unsetAttribute = function unsetAttribute(name, nullif
+  assert(typeof options === 'object', 'The options argument must be an object');
+  assert(typeof cb === 'function', 'The cb argument must be a function');
+
+  var inst = this;
+  var Model = this.constructor;
+  var connector = inst.getConnector();
+  assert(typeof connector.replaceAttributes === 'function',
+    'replaceAttributes() must be implemented by the connector');
+
+  if (isPKMissing(Model, cb))
+    return cb.promise;
+
+  var allowExtendedOperators = connector.settings
+    && connector.settings.allowExtendedOperators;
+
+  var strict = this.__strict;
+  var model = Model.modelName;
+  var hookState = {};
+
+  // Convert the data to be plain object so that update won't be confused
+  if (data instanceof Model) {
+    data = data.toObject(false);
+  }
+  data = removeUndefined(data);
+
+  // Make sure id(s) cannot be changed
+  var idNames = Model.definition.idNames();
+  for (var i = 0, n = idNames.length; i < n; i++) {
+    var idName = idNames[i];
+    if (data[idName] !== undefined && !idEquals(data[idName], inst[idName])) {
+      var err = new Error('id property (' + idName + ') ' +
+        'cannot be updated from ' + inst[idName] + ' to ' + data[idName]);
+      err.statusCode = 400;
+      process.nextTick(function() {
+        cb(err);
+      });
+      return cb.promise;
+    }
+  }
+
+  var context = {
+    Model: Model,
+    where: byIdQuery(Model, getIdValue(Model, inst)).where,
+    data: data,
+    currentInstance: inst,
+    hookState: hookState,
+    options: options
+  };
+
+  Model.notifyObserversOf('before save', context, function(err, ctx) {
+    if (err) return cb(err);
+    data = ctx.data;
+
+    if (strict && !allowExtendedOperators) {
+      var props = Model.definition.properties;
+      var keys = Object.keys(data);
+      var result = {}, key;
+      for (var i = 0; i < keys.length; i++) {
+        key = keys[i];
+        if (props[key]) {
+          result[key] = data[key];
+        } else if (strict === 'throw') {
+          cb(new Error('Unknown property: ' + key));
+          return;
+        } else if (strict === 'validate') {
+          inst.__unknownProperties.push(key);
+        }
+      }
+      data = removeUndefined(result);
+    }
+    
+    // update instance's properties
+    inst.setAttributes(data);
+
+    inst.isValid(function (valid) {
+      if (!valid) {
+        cb(new ValidationError(inst), inst);
+        return;
+      }
+
+      inst.trigger('save', function (saveDone) {
+        inst.trigger('update', function (done) {
+          var typedData = {};
+
+          for (var key in data) {
+            // Convert the properties by type
+            inst[key] = data[key];
+            typedData[key] = inst[key];
+            if (typeof typedData[key] === 'object'
+              && typedData[key] !== null
+              && typeof typedData[key].toObject === 'function') {
+              typedData[key] = typedData[key].toObject();
+            }
+          }
```

Comment:
I assume this was copied from `updateAttributes`. Could you please extract this code to a shared function called from both places?